src/main/java/com/es/phoneshop/model/product/service/DefaultRecentProductsService.java

```diff
@@ -0,0 +1,73 @@
+package com.es.phoneshop.model.product.service;
+
+import com.es.phoneshop.model.cart.service.DefaultCartService;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.dao.ArrayListProductDao;
+import com.es.phoneshop.model.product.dao.ProductDao;
+import jakarta.servlet.http.HttpSession;
+
+import java.util.*;
+
+public class DefaultRecentProductsService implements RecentProductsService {
+    private static final String RECENT_PRODUCTS_SESSION_ATTRIBUTE =
+            DefaultRecentProductsService.class.getName() + ".recentProducts";
+    private static final int RECENT_PRODUCTS_MAX_COUNT = 3;
+
+    private static volatile RecentProductsService instance;
+    private final ProductDao productDao;
+
+    private DefaultRecentProductsService() {
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    public static RecentProductsService getInstance() {
+        if (instance == null) {
+            synchronized (DefaultCartService.class) {
+                if (instance == null) {
+                    instance = new DefaultRecentProductsService();
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public Queue<Product> getRecentProducts(HttpSession currentSession) {
+        synchronized (currentSession) {
+            Queue<Product> recentProducts = getRecentProductSessionAttribute(currentSession);
+            setRecentProductSessionAttribute(currentSession, recentProducts);
+            return recentProducts;
+        }
+    }
+
+    @Override
+    public void addToRecentProducts(HttpSession currentSession, Long productId) {
+        Objects.requireNonNull(productId);
+
+        synchronized (currentSession) {
+            Queue<Product> recentProducts = getRecentProductSessionAttribute(currentSession);
+            deleteFromRecentProducts(recentProducts, productId);
+            recentProducts.offer(productDao.getProduct(productId));
+            setRecentProductSessionAttribute(currentSession, recentProducts);
+        }
+    }
+
+    private void deleteFromRecentProducts(Queue<Product> recentProducts, Long productId) {
+        recentProducts.stream().filter(currentProduct -> productId.equals(currentProduct.getId())).findAny().ifPresent(recentProducts::remove);
+        if (recentProducts.size() >= RECENT_PRODUCTS_MAX_COUNT) {
+            recentProducts.remove();
+        }
+    }
+
+    private void setRecentProductSessionAttribute(HttpSession currentSession, Queue<Product> recentProducts) {
+        currentSession.setAttribute(RECENT_PRODUCTS_SESSION_ATTRIBUTE, recentProducts);
+    }
+
+    private Queue<Product> getRecentProductSessionAttribute(HttpSession currentSession) {
+        Queue<Product> recentProducts = (LinkedList) currentSession.getAttribute(RECENT_PRODUCTS_SESSION_ATTRIBUTE);
```

Comment:
the name of the method is untrue as you not just retrieve products from session but also create the LinkedList if they are absent